### PR TITLE
Deduper p1

### PR DESCRIPTION
**Pull request by nzavoshy**
Ignore the .html doc. It isn't the most recent rendering but the .rmd should be up-to-date

### CODE REVIEW

File: BI624_deduper_P1.Rmd [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+- initiate an empty dictionary
+- set a variable called previous chromsome to 1 so that we can compare chromosome values later
+- loop through each line of the sam file
+    - seperate the sam file into parts and save each of the following 4 components: UMI, strand value(0 or 16), chromsome and start position with soft 
+    - look at the cigar string
+    - if there has been soft clipping, then determine by how much and adjust the start position accordingly and overwrite this value
+    - check if the UMI is in the list of known UMIs, if it is then continue on
+          - check if the chromsome is equal to the previous value and if it is then continue on
+          - set current chromosome value equal to past chromosome value
+          - generate a tuple of this information
```
Comment (Devin-D) [inline BI624_deduper_P1.Rmd:34]: do you maybe want to put the position info in this tuple? what if you have the the same umi on the same chromosome but at a different position.  maybe thats what you meant by generate tuple of this information but it is not clear.

**Review by Devin-D — COMMENTED** [review]
good job. you have a good grasp of why removing pcr duplicates could be important. your pseudo could use a little more detail in parts. i am sure you know what you mean but it was hard to decipher what you meant from someone outside looking in. maybe think about how you could use higher level functions to do a lot of the work and save you some run time. also think about if you want to keep a read that has been softclipped vs not and how that effects your algorithm. overall great job.

File: BI624_deduper_P1.Rmd (hunk 1 above)
Comment (lnassar) [reply BI624_deduper_P1.Rmd:34]: As I understand she's doing that, the tuple is comprised of the 4 components: UMI, strand value, chromosome, start position.

File: BI624_deduper_P1.Rmd [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+          - check if the tuple is in the dictionary, if not then add it to the dictionary with a count of 1 and write that line to an output file * the counter here acts as a sanity check, since if the dictionary is printed and there are numbers >1, then there is a bug in the code
+          - if it isnt then move on to the next line and do nothing else to this line
+
+
```
Comment (lnassar) [inline BI624_deduper_P1.Rmd:38]: The script looks solid. You're employing an approach very similar to mine. (As I'm sure alot of us are).  The only thing I'd mention is you're going to have a dictionary with Tons of keys and really no values. I used the UMI's as the keys and then a list within each for your '4 components'. Just makes it a little easier to follow. The only thing is you wouldn't be able to have your sanity check =( . Also I'd specify exactly which parts of your script you'd want to make functions; cigar string parsing, soft clip fixing, etc. 

In any case looks great, just throwing some ideas at you!

**Pull request merged**